r/R/trajectory_analysis.R

```diff
@@ -0,0 +1,127 @@
+#'
+#' Pseudotime is a cell value that shows the relative progression of the cell in time for a given biological process.
+#' These values will be used to plot the cells along a continuous path that represents the evolution of the process.
+#' This plot represents the final plot of the trajectory analysis, which shows a map of how cells “differentiate” through time, starting from cells in the earlier pseudotime growing into cells in later pseudotime, following the trajectory.
+#'
+#' @param req {body: {
+#'               rootNodes: root nodes ids. Determines the root nodes of the trajectory
+#'              }
+#'            }
+#' @param data SeuratObject
+#'
+#' @return a tibble with pseudotime values
+#' @export
+runTrajectoryAnalysis <- function(req, data) {
+  root_nodes <- req$body$rootNodes
+
+  cell_data <- generateGraphData(data)
+  cell_data <- monocle3::order_cells(cell_data, reduction_method = "UMAP", root_pr_nodes = root_nodes)
+
+  pseudotime <- as.data.frame(cell_data@principal_graph_aux@listData$UMAP$pseudotime)
+
+  # fill in the NULL values for filtered cells
+  pseudotime <- fillNullForFilteredCells(pseudotime, data)
+
+  return(unname(pseudotime))
+}
+
+
+#' Convert Seurat object to Monocle3 cell_data_set object, cluster cells and learn graph
+#'
+#' In order to perform the trajectory analysis with Monocle3, the Seurat object needs to be converted to a Monocle3 cell_data_set object.
+#' This function calculates also embeddings and learn the graph.
+#'
+#' @param data Seurat object
+#'
+#' @return a cell_data_set object with cluster and graph information stored internally
+#' @export
+generateGraphData <- function(data) {
+  cell_data <- SeuratWrappers::as.cell_data_set(data)
+
+  set.seed(ULTIMATE_SEED)
+
+  cell_data <- monocle3::cluster_cells(cds = cell_data, reduction_method = "UMAP")
+  cell_data <- monocle3::learn_graph(cell_data, use_partition = TRUE)
+
+  return(cell_data)
+}
+
+
+#' Fill in the NULL values for filtered cells
+#'
+#' When a dataframe with cell barcodes and some associated information is generated from a filtered Seurat object, it will not contain the cell barcodes that were filtered out.
+#' However, the UI needs the whole array of ordered and unfiltered cell ids.
+#' For this reason, there is a need to add to the dataframe the corresponding cell ids for cell barcodes that were filtered out and fill the values with NULL.
+#'
+#' @param df a dataframe with barcodes as rownames
+#' @param data Seurat object with cells id stored in: data@@meta.data$cells_id
+#'
+#' @return a tibble filled with NULL values for missing cell ids
+#' @export
+fillNullForFilteredCells <- function(df, data) {
+  # add cells_id column
+  df$cells_id <- data@meta.data$cells_id
+  # get max value for cells_id
+  max_value <- max(data@meta.data$cells_id)
+  # order by cells_id
+  df <- df[order(df$cells_id), ]
+  # add NULL values for filtered cells and remove cells_id column
+  df <- df %>%
```

Comment:
```suggestion
  df$cells_id <- data@meta.data$cells_id
  max_value <- max(data@meta.data$cells_id)
  df <- df[order(df$cells_id), ]
  
  df <- df %>%
```

Reply:
These comments are not necessary. the code is pretty self explanatory! And is also clearer thanks to the function docs